### PR TITLE
Re-structure document ID generation favoring `_id` inverted index compression

### PR DESCRIPTION
**Pull request by salvatore-campagna**
This implementation restructures auto-generated document IDs to maximize compression within Lucene's terms dictionary. The key insight is placing stable or slowly-changing components at the start of the ID - the most significant bytes of the timestamp change very gradually (the first byte shifts only every 35 years, the second every 50 days). This careful ordering means that large sequences of IDs generated close in time will share common prefixes, allowing Lucene's Finite State Transducer (FST) to store terms more compactly.

To maintain uniqueness while preserving these compression benefits, the ID combines three elements: a timestamp that ensures time-based ordering, the coordinator's MAC address for cluster-wide uniqueness, and a sequence number for handling high-throughput scenarios. The timestamp handling is particularly robust, using atomic operations to prevent backwards movement even if the system clock shifts.

For high-volume indices generating millions of documents, this optimization can lead to substantial storage savings while maintaining strict guarantees about ID uniqueness and ordering.

**Duplicate ID Prevention**
A duplicate ID could theoretically only occur under the following conditions:

* Two different nodes would need to have identical MAC addresses (definitely not something to expect in normal scenarios)
* These nodes would need to generate IDs in the exact same millisecond
* These nodes would need to generate the exact same sequence number (from the 16.7M available per millisecond)

The implementation prevents this through several mechanisms:

* The MAC address ensures node-level uniqueness, even if timestamps collide
* The atomic sequence counter provides 24 bits (16.7M) of unique IDs per millisecond per node - this means a single coordinator would need to sustain a throughput of 16.7 billion requests per second (16.7M/ms) for the sequence counter to overflow. This is far beyond current hardware capabilities, as it would require processing one document every ~60 picoseconds
* If the sequence counter overflows within a millisecond, the timestamp is automatically incremented
* The atomic timestamp handling ensures monotonic increases even during system clock adjustments

In practice, this means duplication is prevented unless there's a hardware-level violation of MAC address uniqueness, making the IDs safe for distributed systems.

### CODE REVIEW

**Comment by salvatore-campagna** [comment]
This is a draft PR for the Spacetime project https://github.com/elastic/elasticsearch-team/issues/1389

File: server/src/main/java/org/elasticsearch/common/ReverseTimeBasedUUIDGenerator.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        assert macAddress.length == 6;
+        System.arraycopy(macAddress, 0, uuidBytes, i, macAddress.length);
+        i += macAddress.length;
+
+        // From hereinafter everything is almost like random and does not compress well
+        // due to unlikely prefix-sharing
+        uuidBytes[i++] = (byte) (timestamp >>> 8);
+        uuidBytes[i++] = (byte) (sequenceId >>> 8);
+        uuidBytes[i++] = (byte) timestamp;
+        uuidBytes[i++] = (byte) (sequenceId >>> 16);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/ReverseTimeBasedUUIDGenerator.java:48]: This byte only changes once every 65k indexed documents, maybe it's worth moving closer to the front, e.g. right after the MAC address?

**Review by jpountz — COMMENTED** [review]
Very cool. The current format tries to optimize for terms dictionary lookups, which is probably not the best trade-off given that auto-generated IDs are generally coupled with optimizations to skip terms dictionary lookups. Since the last time when we touched the format of auto-generated IDs, Lucene also optimized ID lookups to skip any work if the ID to look up is not in the range of values handled by the segment (https://github.com/apache/lucene/issues/10023) so there is value in having IDs roughly ordered by ingestion timestamp, like what you're doing here.

Let's handle the backward compatibility logic and ship it?

File: docs/changelog/104683.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+pr: 104683
+summary: "Feature: re-arrange document id bytes to take advantage of shared prefixes"
+area: TSDB
```
Comment (jpountz) [inline docs/changelog/104683.yaml:3]: Is it the right area, I thought it would not apply to TSDB since TSDB uses its own IDs and not auto-generated IDs? (I may be outdated, feel free to correct me!)

Comment (salvatore-campagna) [reply docs/changelog/104683.yaml:3]: Should have used the other label indeed :StorageEngine/Logs...definitely not TSDB...TSDB does not use this scheme for doc _id generation.

File: server/src/main/java/org/elasticsearch/common/ReverseTimeBasedUUIDGenerator.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        uuidBytes[i++] = (byte) (timestamp >>> 24); // changes every ~4.5h
+        uuidBytes[i++] = (byte) (timestamp >>> 16); // changes every ~65 secs
+
+        // MAC address of the coordinator might change if there are many coordinators in the cluster
+        // and the indexing api does not necessarily target the same coordinator.
+        byte[] macAddress = macAddress();
+        assert macAddress.length == 6;
+        System.arraycopy(macAddress, 0, uuidBytes, i, macAddress.length);
+        i += macAddress.length;
+        uuidBytes[i++] = (byte) (sequenceId >>> 16);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/ReverseTimeBasedUUIDGenerator.java:42]:
```suggestion
        uuidBytes[i++] = (byte) (sequenceId >>> 16); // changes every ~65k ingested docs
```

**Comment by salvatore-campagna** [comment]
I did the change moving the most significant bits of the sequence number further up, just after the mac address and I see 4% total storage saving with 4.8 GB less space taken and a bit more than 80% saving on individual document _id fields.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java [hunk 4]
```diff
@@ -147,11 +149,17 @@ public void checkIndexSplitAllowed() {}
...
+        private final IndexVersion creationVersion;
+        private final String dataStreamName;
+        private final IndexMode indexMode;
 
         IdAndRoutingOnly(IndexMetadata metadata) {
             super(metadata);
+            this.creationVersion = metadata.getCreationVersion();
             MappingMetadata mapping = metadata.mapping();
             this.routingRequired = mapping == null ? false : mapping.routingRequired();
+            this.dataStreamName = metadata.getDataStreamName();
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:161]: Will probably replace with a boolean.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 5]
```diff
@@ -517,7 +517,8 @@ private ClusterState applyCreateIndexWithTemporaryService(
                     temporaryIndexMeta.getRoutingNumShards(),
                     sourceMetadata,
                     temporaryIndexMeta.isSystem(),
-                    currentState.getMinTransportVersion()
+                    currentState.getMinTransportVersion(),
+                    request.dataStreamName() != null
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:521]: I see in other places in `MetadataCreateIndexService` we check if `dataStreamName` is not `null` to check if   we are dealing with a data stream or not.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java [hunk 6]
```diff
@@ -648,6 +648,7 @@ public Iterator<Setting<?>> settings() {
     private final Double writeLoadForecast;
     @Nullable
     private final Long shardSizeInBytesForecast;
+    private final boolean isDataStream;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:651]: I remember in the past there was push back for also recording in `IndexMetadata` that an index is part of a data stream. Right now this can be derived from `DataStream`. Indices can be moved out of a data stream and then this needs to be updated as well. 


Thinking more about this, I think we don't need to this change. I think we should make the decision on whether to use the new uuid generator based on index mode only.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java [hunk 7]
```diff
@@ -147,11 +149,17 @@ public void checkIndexSplitAllowed() {}
 
     private abstract static class IdAndRoutingOnly extends IndexRouting {
         private final boolean routingRequired;
+        private final IndexVersion creationVersion;
+        private final boolean isDataStream;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:153]: Like I mentioned let's for now only use index mode as a heuristic to use the new uuid generator.

**Review by martijnvg — COMMENTED** [review]
Thanks for updating this PR @salvatore-campagna.

I left two comments around the heuristic that determines when to use new uuid generator for generating the _id. I think we should only based on whether the index mode is logsdb. This should avoid making changes to `IndexMetadata` and `MetadataCreateIndexService`.  It think whether the index mode is logsdb is a sufficient condition. We know if index mode is logsdb, then the use case is logging. No need to also check whether an index is part of a data stream.

I think enabling this new uuid generator just for logsdb is safe, since we know the use case, so we don't affect use cases that rely on faster by id lookups when the _id gets generated.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java (hunk 6 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java]: Do we know the reason behind the push back? Is it because `IndexMetadata` is more about indices rather then data streams? Or maybe because the status would not reflect reality in case the index is taken out of the data stream?

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java]: Yes, because we would need to maintain in two places whether an index is part of a data stream. In `DataStream` itself and in `IndexMetadata`.  And this increases the chances of potential bugs.

For example you need to think about restoring only a backing index from a snapshot or the modify data stream api. The changes in the PR didn't take this into account, because according to the `isDataStream` property an index would be part of a data stream even if no data stream referred to this index.

File: docs/changelog/104683.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+pr: 104683
+summary: "Feature: re-arrange document id bytes to take advantage of shared prefixes"
```
Comment (martijnvg) [inline docs/changelog/104683.yaml:2]: Perhaps rephrase this to emphasize that this reduces storage usage for the _id field? A trade off that works well for logging use cases?

File: server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+ * <p>
+ * <b>Timestamp:</b> Represents the current time in milliseconds, ensuring ordering and uniqueness.
+ * <br>
+ * <b>MAC Address:</b> Ensures uniqueness across different coordinators.
+ * <br>
+ * <b>Sequence ID:</b> Differentiates UUIDs generated within the same millisecond, ensuring uniqueness even at high throughput.
+ * <p>
+ * The result is a compact base64-encoded string, optimized for efficient compression of the _id field in an inverted index.
+ */
+public class TimeBasedKOrderedUUIDGenerator extends TimeBasedUUIDGenerator {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java:30]: Maybe add similar unit tests for this new uuid generator as exists for existing uuid generators in `UUIDTests`?

**Review by martijnvg — COMMENTED** [review]
This looks good! I left two small comments.

File: server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java (hunk 9 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java:30]: I will just add it to the existing `UUIDTests`.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java:30]: that sounds good to me 👍

**Review by martijnvg — APPROVED** [review]
LGTM

I think we should also backport this to the 8.x branch?

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/116810)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
